src/neps/api.py

```diff
@@ -198,9 +200,134 @@ def run(
         )
         max_cost_total = searcher_kwargs["budget"]
         del searcher_kwargs["budget"]
-
+    
     logger = logging.getLogger("neps")
     logger.info(f"Starting neps.run using root directory {root_directory}")
+    
+    if isinstance(searcher, BaseOptimizer):
+        searcher_instance = searcher
+        searcher_alg = "custom"
+        user_defined_searcher = True
+    else:
+        (
+            searcher_instance, 
+            searcher_alg, 
+            searcher_config, 
+            searcher_info, 
+            user_defined_searcher
+        ) = _run_args(
+            run_pipeline=run_pipeline,
+            pipeline_space=pipeline_space,
+            overwrite_working_directory=overwrite_working_directory,
+            development_stage_id=development_stage_id,
+            task_id=task_id,
+            max_evaluations_total=max_evaluations_total,
+            max_evaluations_per_run=max_evaluations_per_run,
+            continue_until_max_evaluation_completed=continue_until_max_evaluation_completed,
+            max_cost_total=max_cost_total,
+            ignore_errors=ignore_errors,
+            loss_value_on_error=loss_value_on_error,
+            cost_value_on_error=cost_value_on_error,
+            pre_load_hooks=pre_load_hooks,
+            logger=logger,
+            searcher=searcher,
+            searcher_path=searcher_path,
+            **searcher_kwargs,
+        )
+
+    # Used to create the yaml holding information about the searcher.
+    # Also important for testing and debugging the api.
+    searcher_info = {
+        "searcher_name": searcher,
```

Comment:
Serializing 'searcher_info' which includes the 'searcher_name' as an instance, will lead to an error during YAML serialization in metahyper for ".optimizer_info.yaml". This error typically takes the form of a TypeError, signifying that the object is not JSON-serializable.
 
To resolve this issue, a quick workaround is to change the 'searcher_name' to a string during initialization:
```python:
    searcher_info = {
        "searcher_name": str(searcher), 
        "searcher_alg": searcher_alg,
        "user_defined_searcher": user_defined_searcher, 
        "searcher_args_user_modified": False,   
}
```
Keep in mind that this approach will result in 'searcher_name' being serialized as (assuming BayesianOptimization):
<neps.optimizers.bayesian_optimization.optimizer.BayesianOptimization object at 0xHEX...>
It's essential to consider alternative solutions for improving the serialization process.